docs/libvirt-pod-networking.md

```diff
@@ -0,0 +1,82 @@
+# Using CNI Interfaces and libvirt for networking
+
+Author: Fabian Deutsch \<fabiand@redhat.com\>
+
+## Introduction
+
+Today networking in KubeVirt is not integrated with Kubernetes, a user can
+rather choose to create a specific VM Spec definition to create a particular
+libvirt domxml, to reuse an existing physical device.
+But this is obviously suboptimal
+
+- It is not integrated with Kubernetes
+- It does not provide connectivity with pods
+- It requires to specify host specific bits (the network device to use)
+
+The purpose of this document is to describe one approach of how we can provide
+a minimal integrated solution, with few workarounds.
+
+
+## Use-case
+
+The primary use-case is to allow VM to get connected to the pod network.
+
+
+## API
+
+The API is pretty clear: As usual an interface will be defined.
+To indicate that Pod interfaces shall be used for networking, a new type
+`pod-network` is introduced.
+All other fields, concerning the properties of the virtual NIC, can be used
+as usual.
+
+```yaml
+kind: VM
+spec:
+  domain:
+    devices:
+      interfaces:
+      - type: pod-network
+        mac:
+          address: 00:11:22:33:44:55
+        model:
+          type: virtio
+        source:
+          network: default
+```
+
+The `source` field can be omitted initially, as Kubernetes only provides a
+single network for Pods today.
+In future we might want to allow it to select the Pod network to connect to.
+
+
+## Implementation
+
+The broader steps to enable networking in the described way:
+
+1. Request/Create a new Pod interface on the libvirt pod
+2. Remember the IP of the Pod interface, then remove it from the Pod interface
+3. Create a libvirt network, backed by a bridge, enable DHCP,
+   and add a single host entry, with the IP address of the Pod, and the MAC
+   address of the corresponding virtual NIC of the VM
```

Comment:
There's a couple of other techniques we should examine in this doc (even if only to eliminate them from consideration). 

First, would be to simply move the eth0 from the VM's POD into the libvirtd POD. This quite trivial - it just requires a rename of the device & moving its network namespace to that associated with the libvirtd POD. We can now give this NIC straight to QEMU for use. This kind of approach would avoid the problems with any NIC anti-spoofing filters, as those are typically associated with network switch (virtual) ports, and since we're using the original those still apply normally.  A potential problem with this is that it may well upset kubelet for the NIC it owns to be moved to a different namespace.

A second approach would be to attach a macvtap device to the eth0 in the VM's POD. Then move this macvtap device into the libvirtd POD and given to QEMU.  This is basically the same as the previous technique, but we've avoided moving the original NIC associated with the VM POD, instead we've attached a slave to it.  The potential problem with this is whether it is possible to attach a macvtap device to every possible type of NIC that might be used in the VM POD.  eg if the NIC visible in the VM POD is alrady a macvtap device, can you chain a macvtap device onto another macvtap device ? I've no idea.